rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalWorkspaceService.java

```diff
@@ -0,0 +1,93 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.rascalmpl.vscode.lsp.rascal;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.eclipse.lsp4j.ClientCapabilities;
+import org.eclipse.lsp4j.FileOperationFilter;
+import org.eclipse.lsp4j.FileOperationOptions;
+import org.eclipse.lsp4j.FileOperationPattern;
+import org.eclipse.lsp4j.FileOperationsServerCapabilities;
+import org.eclipse.lsp4j.RenameFilesParams;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.WorkspaceFolder;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.rascalmpl.vscode.lsp.BaseWorkspaceService;
+import org.rascalmpl.vscode.lsp.IBaseTextDocumentService;
+import org.rascalmpl.vscode.lsp.util.locations.Locations;
+
+public class RascalWorkspaceService extends BaseWorkspaceService {
+    private static final Logger logger = LogManager.getLogger(RascalWorkspaceService.class);
+
+    private final IBaseTextDocumentService docService;
+    private @MonotonicNonNull LanguageClient client;
+
+    RascalWorkspaceService(ExecutorService exec, IBaseTextDocumentService documentService) {
+        super(exec, documentService);
+        this.docService = documentService;
+    }
+
+    @Override
+    public void initialize(ClientCapabilities clientCap, @Nullable List<WorkspaceFolder> currentWorkspaceFolders,
+            ServerCapabilities capabilities) {
+        super.initialize(clientCap, currentWorkspaceFolders, capabilities);
+
+        var fileCap = new FileOperationsServerCapabilities();
+        fileCap.setDidRename(new FileOperationOptions(
+            List.of(new FileOperationFilter(new FileOperationPattern("**")))
+        ));
+
+        capabilities.getWorkspace().setFileOperations(fileCap);
+    }
+
+    @Override
+    public void connect(LanguageClient client) {
+        super.connect(client);
+        this.client = client;
+    }
+
+    @Override
+    public void didRenameFiles(RenameFilesParams params) {
+        logger.debug("workspace/didRenameFiles: {}", params.getFiles());
+
+        CompletableFuture.supplyAsync(() ->
+            workspaceFolders()
+                .stream()
+                .map(f -> Locations.toLoc(f.getUri()))
+                .collect(Collectors.toSet())
+            , getExecuter()
+        ).thenAccept(folders -> ((RascalTextDocumentService) docService).didRenameFiles(params, folders));
```

Comment:
nit: we should move this cast to the constructor, and then throw an exception if we get a non `RascalTextDocumentService` as argument.

Reply:
Since we now delegate to a function of the interface, we don't need this case at all. But perhaps, we should still check the instance type in the constructor?

Reply:
well, if we don't need a cast, let's not cast indeed :) and don't add more requirements than needed, it'll only hurts us later.